pySDC/projects/DAE/sweepers/RungeKuttaDAE.py

```diff
@@ -0,0 +1,181 @@
+    CrankNicholson,
+    EDIRK4,
+    DIRK43_2,
+)
+
+
+class RungeKuttaDAE(RungeKutta):
+    r"""
+    Custom sweeper class to implement Runge-Kutta (RK) methods for general differential-algebraic equations (DAEs)
+    of the form
+
+    .. math::
+        0 = F(u, u', t).
+    
+    RK methods for general DAEs have the form
+
+    .. math::
+        0 = F(u_0 + \Delta t \sum_{j=1}^M a_{i,j} U_j, U_m),
+
+    .. math::
+        u_M = u_0 + \Delta t \sum_{j=1}^M b_j U_j.
+
+    In pySDC, RK methods are implemented in the way that the coefficient matrix :math:`A` in the Butcher
+    tableau is a lower triangular matrix so that the stages are updated node-by-node. This class therefore only supports
+    RK methods with lower triangular matrices in the tableau.
+
+    Parameters
+    ----------
+    params : dict
+        Parameters passed to the sweeper.
+
+    Attributes
+    ----------
+    du_init : dtype_f
+        Stores the initial condition for each step.
+
+    Note
+    ----
+    When using a RK sweeper to simulate a problem make sure the DAE problem class has a ``du_exact`` method since RK methods need an initial
+    condition for :math:`u'(t_0)` as well.
+
+    In order to implement a new RK method for DAEs a new tableau can be added in ``pySDC.implementations.sweeper_classes.Runge_Kutta.py``.
+    For example, a new method called ``newRungeKuttaMethod`` with nodes :math:`c=(c_1, c_2, c_3)`, weights :math:`b=(b_1, b_2, b_3)` and
+    coefficient matrix
+
+    ..math::
+        \begin{eqnarray}
+            A = \begin{pmatrix}
+                a_{11} & 0 & 0 \\
+                a_{21} & a_{22} & 0 \\
+                a_{31} & a_{32} & & 0 \\
+            \end{pmatrix}
+        \end{eqnarray}
+
+    can be implemented as follows:
+
+    >>> class newRungeKuttaMethod(RungeKutta):
+    >>>     nodes = np.array([c1, c2, c3])
+    >>>     weights = np.array([b1, b2, b3])
+    >>>     matrix = np.zeros((3, 3))
+    >>>     matrix[0, 0] = a11
+    >>>     matrix[1, :2] = [a21, a22]
+    >>>     matrix[2, :] = [a31, a32, a33]
+    >>>     ButcherTableauClass = ButcherTableau
+
+    The new class ``newRungeKuttaMethodDAE`` can then be used by defining the DAE class inheriting from both, this base class and class containing
+    the Butcher tableau:
+
+    >>> class newRungeKuttaMethodDAE(RungeKuttaDAE, newRungeKuttaMethod):
+    >>>     pass
+
+    More details can be found [here](https://github.com/Parallel-in-Time/pySDC/blob/master/pySDC/implementations/sweeper_classes/Runge_Kutta.py).
+    """
+
+    def __init__(self, params):
+        super().__init__(params)
+        self.du_init = None
+
+    def predict(self):
+        """
+        Predictor to fill values at nodes before first sweep.
+        """
+
+        # get current level and problem
+        lvl = self.level
+        prob = lvl.prob
+
+        du_init = self.du_init[:] if self.du_init is not None else prob.du_exact(lvl.time)[:]
+        lvl.f[0] = prob.dtype_f(du_init)
```

Comment:
I am a bit irked by the exact solution appearing in the sweeper. But I don't know how to fix it because I am still confused by DAEs. 

We should definitely make sure that the exact solution in the problem class is only called in the first time step. And I would like to see this explicitly in this function. Can we do that somehow? What if we put this in the `__init__` function of this sweeper? Is the problem class properly initialised at that time? 

Or maybe we should make this a parameter of the sweeper. Because you are not limited to the `u_exact` function of the problem for choosing initial conditions. And if you put in something else, this will be totally off. But then the problem is not initialised at the time that you pass the sweeper parameters to the controller during instantiation.

In short: I am not happy, but I don't know a good solution and I am willing to compromise.

Reply:
I'm also not really happy with that. But I haven't found a better solution for that.. Your ideas are quite good. I'd like to realise your first idea. However, as far as I know the level is not yet instantiated at a level when the sweeper is initialised, right? Therefore, calling ``self.du_exact(lvl.time)`` in ``__init__`` would not work?

Reply:
Yes, you are right. How about adding sth like `self.fully_initialized = False` to `__init__` of sweeper. Then you can do in `predict`:
```
if self.fully_initialized:
    du_init = self.du_init[:]
else:
    du_init = prob.du_exact(lvl.time)[:]
    self.fully_initialized = True
```
or sth similar.

Still not sure if this is the best solution, though.

Reply:
I implemented such a statement where we initialize only in the first step (in the first call of this if-else-statement) with the exact solution of ``du``. I still haven't a better solution as well, but for the first time this would be okay and we can move brainstorming to anytime else.